### DISCUSSION TITLE
use directive not found

### DISCUSSION THREAD

**Discussion by jcalfee** [solidjs/solid#1876] [General]
Unless I reference `onClickTheme` in the code outside of the JSX, it appears that the reference is removed.  

### broken:
![image](https://github.com/solidjs/solid/assets/204121/757fc399-8370-4af2-81e9-30a5b8a21ece)

error: `Uncaught ReferenceError: onClickTheme is not defined`

### working:
![image](https://github.com/solidjs/solid/assets/204121/aa8d110b-10e6-4e45-b64e-f9dc77a62e60)

### original:
```js
import { onClickTheme, theme } from "../directives/theme"
import BsToggle2Off from '../icons/BsToggles2Off'
import BsToggle2On from '../icons/BsToggles2On'

// onClickTheme;

export function ThemeButton() {
  return (
    <div use:onClickTheme>
      {theme() === 'light' ? <BsToggle2Off /> : <BsToggle2On/>}
    </div>
  )
}
```

### stack:

```js
import { defineConfig } from 'vite';
import solidPlugin from 'vite-plugin-solid';
import devtools from 'solid-devtools/vite';

export default defineConfig({
  plugins: [
    /*
      https://github.com/thetarnav/solid-devtools/tree/main/packages/extension#readme
    */
    devtools({
      autoname: true,
      locator: {
        targetIDE: 'vscode',
        key: 'Ctrl',
        jsxLocation: true,
        componentLocation: true,
      },
    }),
    solidPlugin(),
  ],
  server: {
    port: process.env.PORT ?? 3037,
    host: process.env.HOST ?? 'localhost'
  },
  build: {
    target: 'esnext',
  }
})
```

**Comment by snnsnn** [comment]
It is a common issue with directives. The imported directive is tree-shaken. To prevent it, you need use the function in the imported module:

```ts
true && directiveFn
```

You can check for a detailed answer: https://stackoverflow.com/a/77049172/7134134

**Comment by ryansolid** [comment]
Ironically its TypeScript not the bundler that is responsible. The bundler gets the code last so it can see it is used. TS on the otherhand gets it first and does its own treeshaking because it needs to remove type only imports.
